fe/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java

```diff
@@ -0,0 +1,259 @@
+    public static final String DYNAMIC_PARTITION_STATE = "dynamicPartitionState";
+    public static final String MSG = "msg";
+
+    private final String DEFAULT_RUNTIME_VALUE = "N/A";
+
+    private Map<String, Map<String, String>> runtimeInfos = Maps.newConcurrentMap();
+    private Set<Pair<Long, Long>> dynamicPartitionTableInfo = Sets.newConcurrentHashSet();
+    private boolean initialize;
+
+    public enum State {
+        NORMAL,
+        ERROR
+    }
+
+
+    public DynamicPartitionScheduler(String name, long intervalMs) {
+        super(name, intervalMs);
+        this.initialize = false;
+    }
+
+    public void registerDynamicPartitionTable(Long dbId, Long tableId) {
+        dynamicPartitionTableInfo.add(new Pair<>(dbId, tableId));
+    }
+
+    public void removeDynamicPartitionTable(Long dbId, Long tableId) {
+        dynamicPartitionTableInfo.remove(new Pair<>(dbId, tableId));
+    }
+
+    public String getRuntimeInfo(String tableName, String key) {
+        Map<String, String> tableRuntimeInfo = runtimeInfos.getOrDefault(tableName, createDefaultRuntimeInfo());
+        return tableRuntimeInfo.getOrDefault(key, DEFAULT_RUNTIME_VALUE);
+    }
+
+    public void removeRuntimeInfo(String tableName) {
+        runtimeInfos.remove(tableName);
+    }
+
+    public void createOrUpdateRuntimeInfo(String tableName, String key, String value) {
+        Map<String, String> runtimeInfo = runtimeInfos.get(tableName);
+        if (runtimeInfo == null) {
+            runtimeInfo = createDefaultRuntimeInfo();
+            runtimeInfo.put(key, value);
+            runtimeInfos.put(tableName, runtimeInfo);
+        } else {
+            runtimeInfo.put(key, value);
+        }
+    }
+
+    private Map<String, String> createDefaultRuntimeInfo() {
+        Map<String, String> defaultRuntimeInfo = Maps.newConcurrentMap();
+        defaultRuntimeInfo.put(LAST_UPDATE_TIME, DEFAULT_RUNTIME_VALUE);
+        defaultRuntimeInfo.put(LAST_SCHEDULER_TIME, DEFAULT_RUNTIME_VALUE);
+        defaultRuntimeInfo.put(DYNAMIC_PARTITION_STATE, State.NORMAL.toString());
+        defaultRuntimeInfo.put(MSG, DEFAULT_RUNTIME_VALUE);
+        return defaultRuntimeInfo;
+    }
+
+    private void dynamicAddPartition() {
+        Iterator<Pair<Long, Long>> iterator = dynamicPartitionTableInfo.iterator();
+        while (iterator.hasNext()) {
+            Pair<Long, Long> tableInfo = iterator.next();
+            Long dbId = tableInfo.first;
+            Long tableId = tableInfo.second;
+            Database db = Catalog.getInstance().getDb(dbId);
+            if (db == null) {
+                iterator.remove();
+                continue;
+            }
+            String tableName;
+            ArrayList<AddPartitionClause> addPartitionClauses = new ArrayList<>();
+            db.readLock();
+            try {
+                // Only OlapTable has DynamicPartitionProperty
+                OlapTable olapTable = (OlapTable) db.getTable(tableId);
```

Comment:
When olapTable's state is not NORMAL, should also skip this round. For example, when table is under SCHEMA CHANGE, add partition will not be allowed.